### DISCUSSION TITLE
Supabase auth set custom cookieOptions

### DISCUSSION THREAD

**Discussion by jogly** [supabase/supabase#3198] [Questions] [answered]
# Feature request

## Is your feature request related to a problem? Please describe.

I would like to set the `domain` attribute of the resulting cookie from `supabase.auth.api.setAuthCookie()`.  

## Describe the solution you'd like

Include gotrue-js 's `CookieOptions` in `createClient(_,_,opts /* ← here */)` to be merged w `constants.COOKIE_OPTIONS` in @supabase/gotrue-js.

## Describe alternatives you've considered

none

## Additional context

none

**Comment by thorwebdev** [comment]
They default cookie options should work for most out of the box. Should you need to modify them, you can initialise the gotrue-js client yourself and pass cookie options to the constructor:

```js
import { GoTrueClient } from '@supabase/gotrue-js'

const myCustomAuthClient = new GoTrueClient({
  url: `${process.env.NEXT_PUBLIC_SUPABASE_URL}/auth/v1`,
  headers: {
    Authorization: `Bearer ${process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY}`,
    apikey: `${process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY}`,
  },
  cookieOptions: {
    name: 'sb:token',
    lifetime: 60 * 60 * 8,
    domain: '',
    path: '/',
    sameSite: 'lax',
  }
})

export default function handler(req, res) {
  myCustomAuthClient.api.setAuthCookie(req, res)
}
```

**Reply by philohelp** [reply]
I tried the code above but strangely it doesn't attach the session to the cookie - resulting in setting an empty cookie. Although I checked that session was indeed sent in the request (& it works with supabase.auth.api.setAuthCookie(req, res)). I tried to check the code for the goTrueClient / goTrueApi and I have no idea why the session would get lost in the process.  

This is the request:

```
  const setServerSession = async (event: AuthChangeEvent, session: Session) => {
    await fetch("/api/auth", {
      method: "POST",
      headers: new Headers({ "Content-Type": "application/json" }),
      credentials: "same-origin",
      body: JSON.stringify({ event, session }),
    });
  };
```

**Reply by thorwebdev** [reply]
@philohelp what's your server-side code?

**Reply by philohelp** [reply]
I think it's quite the same code as above. 

```
const myCustomAuthClient = new GoTrueClient({
  url: "https://SUPA-URL.supabase.co/auth/v1",
   headers: {
     Authorization:
       "Bearer SUPA-KEY",
        "apikey: SUPA-KEY",
   },
   cookieOptions: {
     name: "sb:token",
     lifetime: 60 * 60 * 8,
     domain: "http://welcome.localhost:3000", // "http://USERNAME.localhost:3000" won't work either
     path: "/",
     sameSite: "lax",
   },
});
```

My goal is to be able to specify the subdomain, as I'm working on a mutli tenant app. The users will signin / register from USERNAME.app.com but I can't ask supabase to allow to redirect them directly to this address  because I won't know the username in advance.

So I was trying to follow this path:
1- after signin, redirect the user to a generic address (like welcome.app.com) - which is authorized on the Supabase settings panel
2- set a cookie from there, with USERNAME.app.com specified as the domain
3- redirect the user to USERNAME.app.com where he'll be automatically registered.

**Reply by thorwebdev** [reply]
@philohelp sorry, this slipped through the cracks. Generally just setting the domain should work fine: "For example, if you set Domain=mozilla.org, cookies are available on subdomains like developer.mozilla.org." (https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies). However I believe this doesn't work on localhost.

**Reply by micheltucker** [reply]
> They default cookie options should work for most out of the box. Should you need to modify them, you can initialise the gotrue-js client yourself and pass cookie options to the constructor:
> 
> ```js
> import { GoTrueClient } from '@supabase/gotrue-js'
> 
> const myCustomAuthClient = new GoTrueClient({
>   url: `${process.env.NEXT_PUBLIC_SUPABASE_URL}/auth/v1`,
>   headers: {
>     Authorization: `Bearer ${process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY}`,
>     apikey: `${process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY}`,
>   },
>   cookieOptions: {
>     name: 'sb:token',
>     lifetime: 60 * 60 * 8,
>     domain: '',
>     path: '/',
>     sameSite: 'lax',
>   }
> })
> 
> export default function handler(req, res) {
>   myCustomAuthClient.api.setAuthCookie(req, res)
> }
> ```

Thanks for this. I appreciate that this is quite some time ago. However, I was wondering where I would need to make this change?

**Reply by thorwebdev** [reply]
@micheltucker with supabase-js v2 cookie handling has been removed and should either be approach like this ( https://supabase.com/docs/guides/auth/server-side-rendering ) or via one of the framework helpers: https://github.com/supabase/auth-helpers#supported-frameworks

**Comment by thorwebdev** [comment, marked as answer]
**NOTE**: with supabase-js v2 cookie handling has been removed and should either be approach like this ( https://supabase.com/docs/guides/auth/server-side-rendering ) or via one of the framework helpers: https://github.com/supabase/auth-helpers#supported-frameworks

**Reply by Aldo111** [reply]
Hey I've looked all over but I don't quite understand how we can make this work for subdomains now in nextjs (for multi-tenant applications) if we can't set cookie options like earlier.

 I'm using the middleware and such as recommended by the guides but I can't quite get it to work (eg. when I'm on subdomain.mydomain.com the session doesnt get picked up - neither in the middleware nor in the client and there are no cookies but if i go back to mydomain.com there are cookies)

**Reply by Aldo111** [reply]
Ah disregard my previous comment, I think I misinterpreted your cookie handling comment to mean cookie options no longer works but I see that's not the case since I was able to get it working.

For anyone else looking, I shared what worked for me below (slightly different than the cookie options shown in this ticket earlier)

**Comment by Aldo111** [comment]
For anyone else looking in the future trying to setup a multitenant application with subdomains and trying to get the cookies to work across subdomains locally (requires remapping /etc/hosts to point 127.0.0.1 to localhost.com or something since localhost isnt allowed in cookie options) or in prod, this is what ended up working for me in **NextJS**

middleware.ts (Using nextjs v13 with appdir and the auth nextjs supabase helper library as per the supabase guide)
```ts
const hostname = new URL(process.env.NEXT_PUBLIC_AUTH_DOMAIN).hostname;
  const supabase = createMiddlewareSupabaseClient(
    { req, res },
    {
      cookieOptions: {
        domain: '.' + hostname,
        maxAge: 100000000,
        path: '/',
        sameSite: 'lax',
        secure: process.env.NEXT_PUBLIC_AUTH_DOMAIN.startsWith('https://'),
      },
    }
  );
```

Pretty much the exact same thing in your clientside supabase client

SupabaseProvider.tsx
```tsx
const hostname = new URL(process.env.NEXT_PUBLIC_AUTH_DOMAIN).hostname;
  const [supabase] = useState(() =>
    createBrowserSupabaseClient({
      options: {},
      cookieOptions: {
        domain: '.' + hostname,
        maxAge: 100000000,
        path: '/',
        sameSite: 'lax',
        secure: process.env.NEXT_PUBLIC_AUTH_DOMAIN.startsWith('https://'),
      },
    })
  );
```

**Reply by tomekit** [reply]
On my side after I've applied `cookieOptions`, the `supabase-auth-token` cookie gets duplicated. Using: `@supabase/auth-helpers-nextjs: 0.6.1`

I am running site on: `subdomain.example.com` and would like cookie to be available on: `.example.com`. Using above snippet I set `cookieOptions` for `createBrowserSupabaseClient`, to:
```js
cookieOptions: {
	domain: '.example.com',
	maxAge: 100000000,
	path: '/',
	sameSite: 'lax',
	secure: true,
},
```
Unfortunately when I sign in I got two duplicate cookies set, the old `subdomain.example.com` which is still being set regardless of my change and a new one: `.example.com`. I would expect only new one to appear.

**Reply by Collin0810** [reply]
I got the same issue, have you managed to fix it?

**Reply by TimurBas** [reply]
@tomekit @Collin0810 
Firstly, I have two frontends `example.com` and `app.example.com` and my user logs in on `example.com` the person gets redirected to `app.example.com`. 
Now when the user is redirected to `app.example.com` they have two cookies for some reason.
I managed to remove the one that **is not** defined by the `cookieOptions` that you show. 
Actually I have no idea where it comes from.

The following does the trick for me at least.
Note: This middleware is the middleware for `app.example.com`.

 ```ts
 import { createMiddlewareClient } from "@supabase/auth-helpers-nextjs";
import { NextResponse } from "next/server";

import type { NextRequest } from "next/server";
import { Database } from "supa";

const resolveRedirectUrl = () => {
  const env = process.env.NEXT_PUBLIC_VERCEL_ENV;

  if (env === "production") return process.env.PROD_PUBLIC_FRONTEND_URL;
  if (env === "preview") return process.env.STAGING_PUBLIC_FRONTEND_URL;
  return process.env.LOCAL_PUBLIC_FRONTEND_URL;
};

export async function middleware(req: NextRequest) {
  const res = NextResponse.next();
  const supabase = createMiddlewareClient<Database>({ req, res });
  const {
    data: { session },
  } = await supabase.auth.getSession();

  if (!session)
    return NextResponse.redirect(new URL("/login", resolveRedirectUrl()));

// This part
  if (session?.user) {
    res.cookies.delete("sb-<Supabase-Reference-Id>-auth-token");
    res.cookies.delete("supabase-auth-token");
  }

  return res;
}
 ```
 where `<Supabase-Reference-Id>` is the one you find in the Supabase dashboard under _Project Settings_ -> _General Settings_ -> _Reference ID_.

**Comment by Necmttn** [comment]
A small bug fix for nextjs app dir users;
https://github.com/supabase/auth-helpers/pull/604 
which enables to use subdomain auth
